Automattic/wp-calypso
# composite-checkout: Refactor payment method creation to remove Container

#### Changes proposed in this Pull Request

In #38727 we will be adding a new payment method for full credits payments. 

In order to implement this, some refactoring had to be done, which makes the other PR rather unfortunately large. This PR is the unrelated refactoring only.

Previously, the payment method objects were being created in the `CompositeCheckoutContainer`, which then passed them down to the `CompositeCheckout` component in Calypso to in turn pass them to the `CheckoutProvider` where they are used. However, we only learn about the number of available credits when we download the shopping cart with the `useShoppingCart` hook, and that happens in `CheckoutProvider`. Therefore, this PR moves the payment method creation down to that component.

With that code moved (I moved most of it into a separate file to make it easier to grok since it is getting rather large and there's a lot of caching involved), `CompositeCheckoutContainer` was starting to be rather small, and it seemed to me that there's really no need for it any longer (it was originally present because `CompositeCheckout` was not in the Calypso package, but now it is). Therefore, I moved the remainder of its logic into `CompositeCheckout` and removed the now-unused class. 

This caused some of the tests to fail (for reasons I still don't fully understand) and so I also refactored the tests to make them work again. One of the things I had to do was to allow passing in an array of allowed payment method strings to `CompositeCheckout` to filter the payment methods that would be created. This will be useful later when we need to also use the `allowed_payment_methods` array from the cart endpoint.

Because the payment methods themselves are created before the checkout form is rendered, and two of the payment method rely on async data (credits and existing payment methods), we also needed to refactor when the payment methods get created and how they are cached. It made more sense to me to wait for the payment methods to be fully created before showing the form. Therefore, this PR also treats loading stored cards as a loading state, and moves the payment method caching to the inside of the `CompositeCheckout` component. This allows us to return an empty array of payment methods while things are still loading.

#### Testing instructions

- Run calypso with `ENABLE_FEATURES=composite-checkout-wpcom npm start`.
- Sandbox the store.
- Add a plan to your cart.
- Complete checkout with a card (new or saved) and verify that everything works as expected.
